Lazy initialize producer and defer Kafka log events: #1

aerskine opened the pull request:
Addresses #1.

Also addresses throwing away all Kafka log messages.

Also addresses the undesirable situation in which an appender is configured but not referenced in the logback config -- it will still start a KafkaProducer.

The cost of polling the ConcurrentLinkedQueue used to defer internal Kafka events is about 60 nanos on my machine, which should be acceptable in all cases.

dgomesbr commented:
Hey @aerskine, could you please elaborate on why using commons-lang3? it's a PITA for other platforms that we use here and wouldn't like to include it.

aerskine commented:
Hi - it is just for the LazyInitialize. If it is a PITA I can remove and hand-roll the lazy init.



> On Aug 20, 2015, at 12:07 AM, Diego Magalhães <notifications@github.com> wrote:
> 
> Hey @aerskine, could you please elaborate on why using commons-lang3? it's a PITA for other platforms that we use here and wouldn't like to include it.
> 
> —
> Reply to this email directly or view it on GitHub.
>

danielwegener commented:
Hello @aerskine! Thank you for your PR! I looks very promising at the first glance. I'll take a closer look this evening. I'd  agree with @dgomesbr that I'd like to avoid adding avoidable dependencies :)

aerskine commented:
@dgomesbr @danielwegener Thanks for the feedback!

I removed the commons-lang dependency, but kept the LazyInitializer pattern (with the synchronized init block) rather than use an AtomicReference SO example since we don't want to (potentially) initialize multiple producers.

danielwegener commented:
@aerskine could you rebase+squash this PR please?

aerskine commented:
@danielwegener Rebased!

aerskine commented:
@danielwegener Is there anything further you need from me? Thanks.

danielwegener commented:
Sorry i am on holiday, will report back soon

danielwegener commented:
Hi @aerskine, sorry for putting you off again :( I really had not so much time to look into this. I had a quick look and I am a bit unsure If the MR could still be made a bit more simple (do we need another ConcurrentQueueAppender that has its livecycle not properly managed by logback?). That just were some first impressions and I may have missed some parts whats happening actually, but when I have time I'll give a proper review.

Sorry again for the delay-

aerskine commented:
Hi @danielwegener no worries. You're right of course, the extra appenders were a hangover from an earlier non-lazy attempt I'd made to fix the circular append issue. I've removed them, and used a plain old concurrent queue. I also added a test to ensure kafka message appends are really deferred.

danielwegener commented:
Hell yeah, that looks much better - I like it. Let me think about a few more tests scenarios but overall its LGTM.

danielwegener commented:
Forgot about this one again - sry. Squash it and we're ready to merge!

aerskine commented:
@danielwegener Squashed! Thanks.

danielwegener commented:
@aerskine Merged! Thanks for your contribution and patience :) We still may think about an optimization that works around the concurrent queue lock when the producer is started. The 60 nanos are maybe okay, but contention between all logging threads may not.

aerskine commented:
@danielwegener Thanks! I'll have a think about your contention comment and do some profiling if I get time.
